File: AED2/Trabalhos praticos/TP02/C/main.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#define lineMax 1000
#define auxStringTam 200
#define auxArrayTam 200
#define structList 500
#define attributes 18

typedef struct{
    char stringSplited[attributes][auxStringTam];
} Split;

typedef struct{
    int day;
    int month;
    int year;
} LocalDate;

typedef struct{
    char id[auxStringTam];
    char name[auxStringTam];
    char alternateNames[auxArrayTam];
    char house[auxStringTam];
    char ancestry[auxStringTam];
    char species[auxStringTam];
    char patronus[auxStringTam];
    char hogwartsStaff[auxStringTam];
    char hogwartsStudent[auxStringTam];
    char actorName[auxStringTam];
    char alive[auxStringTam];
    char alternateActors[auxStringTam];
    //char dateOfBirth[auxStringTam];
    // int day;
    // int month;
    // int year;
    LocalDate dateOfBirth;
    int yearOfBirth;
    char eyeColour[auxStringTam];
    char gender[auxStringTam];
    char hairColour[auxStringTam];
    char wizard[auxStringTam];
} Personagem;

void questao01(const Personagem personagens[]);

LocalDate stringDateFormat(char* aux);
void stringBooleanFormat(char* aux);
int stringIntFormat(const char* aux);
void removeStringChar(char* aux, char c);

void readCsv(Personagem personagens[], char* fileToRead);
Personagem searchPessoaId(const Personagem personagem[], char* id);

void printPersonagem(const Personagem* personagem);

int main(){
    Personagem personagens[structList];

    readCsv(personagens, "/tmp/characters.csv");

    questao01(personagens);

    return 0;
}

Personagem searchPessoaId(const Personagem personagens[], char* id){
    Personagem resp;
    strcpy(resp.id, "-1");

    for ( int i = 0; i < structList; i++){
        if ( strcmp(personagens[i].id, id) == 0 ){
            resp = personagens[i];
            break;
        }
    }

    return resp;
}

void questao01(const Personagem personagens[]) {
    char input[auxStringTam];
    Personagem aux;

    scanf(" %s", input);
    while ( (strcmp(input, "FIM") != 0) ){

        aux = searchPessoaId(personagens, input);

        if ( strcmp(aux.id, "-1") != 0 ) printPersonagem(&aux);
        else printf("Id nao encontrado\n");

        scanf(" %s", input);
    }

}

void readCsv(Personagem personagens[], char* fileToRead){
    char aux[lineMax];
    Split split;

    FILE* arq;
    arq = fopen(fileToRead, "r");
    if ( arq == NULL ) printf("Erro ao abrir arquivo");

    fgets(aux, sizeof(aux), arq);

    int k = 0;
    while ( !(feof(arq)) ){
        for ( int i = 0; i < attributes; i++ ) {
            fscanf(arq, "%[^;\n]", split.stringSplited[i]);
            fgetc(arq);
        }

        strcpy(personagens[k].id, split.stringSplited[0]);
        strcpy(personagens[k].name, split.stringSplited[1]);
            removeStringChar(split.stringSplited[2], '[');
            removeStringChar(split.stringSplited[2], ']');
            removeStringChar(split.stringSplited[2], '\'');
        strcpy(personagens[k].alternateNames, split.stringSplited[2]);
        strcpy(personagens[k].house, split.stringSplited[3]);
        strcpy(personagens[k].ancestry, split.stringSplited[4]);
        strcpy(personagens[k].species, split.stringSplited[5]);
        strcpy(personagens[k].patronus, split.stringSplited[6]);
        stringBooleanFormat(split.stringSplited[7]);
        strcpy(personagens[k].hogwartsStaff, split.stringSplited[7]);
            // strcpy(personagens[k].hogwartsStaff, "false");
        stringBooleanFormat(split.stringSplited[8]);
        strcpy(personagens[k].hogwartsStudent, split.stringSplited[8]);
            // strcpy(personagens[k].hogwartsStudent, "false");
        strcpy(personagens[k].actorName, split.stringSplited[9]);
        stringBooleanFormat(split.stringSplited[10]);
        strcpy(personagens[k].alive, split.stringSplited[10]);
            // strcpy(personagens[k].alive, "false");
        strcpy(personagens[k].alternateActors, split.stringSplited[11]);
            // strcpy(personagens[k].dateOfBirth, split.stringSplited[12]);
            // sscanf(split.stringSplited[12], "%d-%d-%d", &personagens[k].day, &personagens[k].month, &personagens[k].year);
        personagens[k].dateOfBirth = stringDateFormat(split.stringSplited[12]);
        personagens[k].yearOfBirth = atoi(split.stringSplited[13]);
        strcpy(personagens[k].eyeColour, split.stringSplited[14]);
        strcpy(personagens[k].gender, split.stringSplited[15]);
        strcpy(personagens[k].hairColour, split.stringSplited[16]);
        stringBooleanFormat(split.stringSplited[17]);
        strcpy(personagens[k].wizard, split.stringSplited[17]);
            // strcpy(personagens[k].wizard, "false");

        for ( int i = 0; i < attributes; i++ ) {
            split.stringSplited[i][0] = 0;
        }

        k++;
    }

    fclose(arq);
}

void stringBooleanFormat(char* aux){
    aux[0] == 'V' ? strcpy(aux, "true") : strcpy(aux, "false");
}

int stringIntFormat(const char* aux){
    return atoi(aux);
}

LocalDate stringDateFormat(char* aux){
    LocalDate formated;

    sscanf(aux, "%d-%d-%d", &formated.day, &formated.month, &formated.year);
    
    return formated;
}

void removeStringChar(char* aux, char c){
    int i = 0;
    int j = 0;
    while (aux[i] != '\0') {
        if (aux[i] != c) aux[j++] = aux[i];
        i++;
    }
    aux[j] = '\0';
}

void printPersonagem(const Personagem* personagem) { //                             LocalDate
    printf("[%s ## %s ## {%s} ## %s ## %s ## %s ## %s ## %s ## %s ## %s ## %s ## %02d-%02d-%02d ## %d ## %s ## %s ## %s ## %s]\n",
            personagem->id,
            personagem->name,
            personagem->alternateNames,
            personagem->house,
            personagem->ancestry,
            personagem->species,
            personagem->patronus,
            personagem->hogwartsStaff,
            personagem->hogwartsStudent,
            personagem->actorName,
            personagem->alive,
            // personagem->day,
            // personagem->month,
            // personagem->year,
            // personagem->dateOfBirth,
            personagem->dateOfBirth.day,
            personagem->dateOfBirth.month,
            personagem->dateOfBirth.year,
            personagem->yearOfBirth,
            personagem->eyeColour,
            personagem->gender,
            personagem->hairColour,
            personagem->wizard);
}

File: AED2/Trabalhos praticos/TP02/Java/Main.java
import java.time.format.DateTimeFormatter;
// import java.io.RandomAccessFile;
import java.io.BufferedReader;
import java.util.ArrayList;
import java.time.LocalDate;
import java.io.FileReader;
import java.util.Scanner;
import java.util.Arrays;
import java.util.List;

public class Main{
    public static void main(String[] args){
        Scanner sc = new Scanner(System.in);

        List<Personagem> personagens = ReadCsv.readAllFile("/tmp/characters.csv");

        questao01(personagens, sc);
        
        sc.close();
    }

    
    static Personagem searchPessoaId(final List<Personagem> personagens, Scanner sc, String id){
        Personagem search = new Personagem();
        search = null;
        
        for ( Personagem pessoa : personagens){
            if ( pessoa.getId().equals(id)){
                search = pessoa;
                break;
            }
        }
        
        return search;
    }
    
    static void questao01(final List<Personagem> personagens, Scanner sc){
        String input = new String();
        Personagem resp = new Personagem();

        while(!(input = sc.nextLine()).equals("FIM")){
            resp = searchPessoaId(personagens, sc, input);

            if ( resp != null ) resp.printPessoa();
            else System.out.println("Id nao encontrado");
        }
    }
}

class Personagem{
    private String id;
    private String name;
    private List<String> alternateNames;
    private String house;
    private String ancestry;
    private String species;
    private String patronus;
    private boolean hogwartsStaff; 
    private boolean hogwartsStudent;
    private String actorName;
    private boolean alive;
    private List<String> alternateActors;
    private LocalDate dateOfBirth;
    private int yearOfBirth;
    private String eyeColour;
    private String gender;
    private String hairColour;
    private boolean wizard;

    public Personagem(String[] aux) {
        this.id = aux[0];
        this.name = aux[1];
        this.alternateNames = new ArrayList<>();
        alternateNames.addAll(Arrays.asList(stringArrayFormat(aux[2])));
        this.house = aux[3];
        this.ancestry = aux[4];
        this.species = aux[5];
        this.patronus = aux[6];
        this.hogwartsStaff = stringBooleanFormat(aux[7]);
        this.hogwartsStudent = stringBooleanFormat(aux[8]);
        //this.hogwartsStaff = false;
        //this.hogwartsStudent = false;
        this.actorName = aux[9];
        this.alive = stringBooleanFormat(aux[10]);
        //this.alive = false;
        this.alternateActors = new ArrayList<>();
        alternateActors.addAll(Arrays.asList(stringArrayFormat(aux[11])));
        this.dateOfBirth = stringLocalDateFormat(aux[12]);
        this.yearOfBirth = Integer.parseInt(aux[13]);
        this.eyeColour = aux[14];
        this.gender = aux[15];
        this.hairColour = aux[16];
        this.wizard = stringBooleanFormat(aux[17]);
        //this.wizard = false;
    }

    public Personagem(){}

    public void setId(String id) {this.id = id;}
    public void setName(String name) {this.name = name;}
    public void setAlternateNames(List<String> alternateNames) {this.alternateNames = alternateNames;}
    public void setHouse(String house) {this.house = house;}
    public void setAncestry(String ancestry) {this.ancestry = ancestry;}
    public void setSpecies(String species) {this.species = species;}
    public void setPatronus(String patronus) {this.patronus = patronus;}
    public void setHogwartsStaff(boolean hogwartsStaff) {this.hogwartsStaff = hogwartsStaff;}
    public void setHogwartsStudent(boolean hogwartsStudent) {this.hogwartsStudent = hogwartsStudent;}
    public void setActorName(String actorName) {this.actorName = actorName;}
    public void setAlive(boolean alive) {this.alive = alive;}
    public void setAlternateActors(List<String> alternateActors) {this.alternateActors = alternateActors;}
    public void setDateOfBirth(LocalDate dateOfBirth) {this.dateOfBirth = dateOfBirth;}
    public void setYearOfBirth(int yearOfBirth) {this.yearOfBirth = yearOfBirth;}
    public void setEyeColour(String eyeColour) {this.eyeColour = eyeColour;}
    public void setGender(String gender) {this.gender = gender;}
    public void setHairColour(String hairColour) {this.hairColour = hairColour;}
    public void setWizard(boolean wizard) {this.wizard = wizard;}

    public String getId() {return id;}
    public String getName() {return name;}
    public List<String> getAlternateNames() {return alternateNames;}
    public String getHouse() {return house;}
    public String getAncestry() {return ancestry;}
    public String getSpecies() {return species;}
    public String getPatronus() {return patronus;}
    public boolean isHogwartsStaff() {return hogwartsStaff;}
    public boolean getHogwartsStudent() {return hogwartsStudent;}
    public String getActorName() {return actorName;}
    public boolean isAlive() {return alive;}
    public List<String> getAlternateActors() {return alternateActors;}
    public LocalDate getDateOfBirth() {return dateOfBirth;}
    public int getYearOfBirth() {return yearOfBirth;}
    public String getEyeColour() {return eyeColour;}
    public String getGender() {return gender;}
    public String getHairColour() {return hairColour;}
    public boolean isWizard() {return wizard;}

    public static Personagem clonePersonagem(Personagem x){
        Personagem clone = new Personagem();

        clone.setId(x.getId());
        clone.setName(x.getName());
        clone.setAlternateNames(new ArrayList<>(x.getAlternateNames()));
        clone.setHouse(x.getHouse());
        clone.setAncestry(x.getAncestry());
        clone.setSpecies(x.getSpecies());
        clone.setPatronus(x.getPatronus());
        clone.setHogwartsStaff(x.isHogwartsStaff());
        clone.setHogwartsStudent(x.getHogwartsStudent());
        clone.setActorName(x.getActorName());
        clone.setAlive(x.isAlive());
        clone.setAlternateActors(new ArrayList<>(x.getAlternateActors()));
        clone.setDateOfBirth(x.getDateOfBirth());
        clone.setYearOfBirth(x.getYearOfBirth());
        clone.setEyeColour(x.getEyeColour());
        clone.setGender(x.getGender());
        clone.setHairColour(x.getHairColour());
        clone.setWizard(x.isWizard());
    
        return clone;
    }

    public void printPessoa() {
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd-MM-yyyy");
        String dateOfBirthFormat = dateOfBirth.format(formatter);

        System.out.println("[" + id + " ## "  + name + " ## " + arrayToStringFormat(alternateNames) + " ## " + house + " ## "
        + ancestry + " ## " + species + " ## " + patronus + " ## " + hogwartsStaff + " ## " + hogwartsStudent + " ## " 
        + actorName + " ## " + alive + " ## " + dateOfBirthFormat + " ## "
        + yearOfBirth + " ## " + eyeColour + " ## " + gender + " ## " + hairColour + " ## " + wizard + "]");
    }

    static String arrayToStringFormat(List<String> aux){
        String[] array = aux.toArray(new String[aux.size()]);
        String str = Arrays.toString(array);

        str = str.replace("[", "{");
        str = str.replace("]", "}");

        return str;
    }
    
    static String[] stringArrayFormat(String aux){
        aux = aux.replace("[", "");
        aux = aux.replace("]", "");
        aux = aux.replace("'","");

        String[] auxArray = aux.split(",");

        for ( int i = 0; i < auxArray.length; i++ ){
            auxArray[i] = auxArray[i].trim();
        }

        return auxArray;
    }

    static boolean stringBooleanFormat(final String aux){
        return aux.equals("VERDADEIRO") ? true : false;
    }

    static LocalDate stringLocalDateFormat(final String aux){
        try {
            DateTimeFormatter formatter = DateTimeFormatter.ofPattern("d-M-yyyy");
            LocalDate data = LocalDate.parse(aux, formatter);
            return data;
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    }

}

class ReadCsv{
    static List<Personagem> readAllFile(final String fileToRead){
        List<Personagem> personagens = new ArrayList<>();
        
        try {
            BufferedReader br = new BufferedReader(new FileReader(fileToRead));
            br.readLine();

            String csvLine = new String();

            while( (csvLine = br.readLine()) != null ){
                String[] csvArray = lineArrayFormat(csvLine, ";");
                Personagem pessoa = new Personagem(csvArray);
                personagens.add(pessoa);
            }

            br.close();
        }
        catch (Exception e){
            e.printStackTrace();
            personagens = null;
        }

        return personagens;
    }


    static String[] lineArrayFormat(String aux, final String c){
        String[] auxArray = aux.split(c);
        return auxArray;
    }

}